keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -26,64 +33,188 @@ class NotificationEventListener(
   @EventListener(PinnedNotification::class)
   fun onPinnedNotification(notification: PinnedNotification) {
     with(notification) {
-      val envNotifications = repository.environmentNotifications(config.name, pin.targetEnvironment)
-      val deliveryArtifact = pin.reference.let {
-        repository.getArtifact(config.name, it)
-      }
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == pin.reference
+      } ?: return
+
       val pinnedArtifact = repository.getArtifactVersion(deliveryArtifact, pin.version, null)
       val currentArtifact = repository.getArtifactVersionByPromotionStatus(config, pin.targetEnvironment, deliveryArtifact, PromotionStatus.CURRENT)
 
-      envNotifications.forEach { notificationConfig ->
-        if (notificationConfig.type == NotificationType.slack) {
-          val handler = handlers.find {
-            it.type == notification.type
-          } as? SlackNotificationHandler<SlackPinnedNotification>
-          handler?.let {
-            it.sendMessage(SlackPinnedNotification(
-              channel = notificationConfig.address,
-              pin = pin,
-              currentArtifact = currentArtifact,
-              pinnedArtifact = pinnedArtifact,
-              application = config.application,
-              time = clock.instant()
-            ))
-          }
+      if (pinnedArtifact == null || currentArtifact == null) {
+        log.debug("can't send notification as either pinned artifact or current artifacts information is missing")
+        return
+      }
+
+      try {
+        config.environments.first { environment ->
+          environment.name == pin.targetEnvironment
+        }.also {
+          it.sendSlackMessage(SlackPinnedNotification(
+            pin = pin,
+            currentArtifact = currentArtifact,
+            pinnedArtifact = pinnedArtifact,
+            application = config.application,
+            time = clock.instant()
+          ),
+            Type.ARTIFACT_PINNED)
         }
+      } catch (ex: NoSuchElementException) {
+        log.debug("no environment ${pin.targetEnvironment} was found in the config named ${config.name}")
+        throw ex
```

Comment:
Does this warrant throwing an exception? Or would it be better to log an error and alert on this? 

(I don't know the answer. I just tend to question throwing exceptions because we don't always handle that nicely downstream)

Reply:
hmm good question. honestly, this whole notification functionality won't really affect the user if something fails, as the user will get an error only if the actual pinning/ unpinning/ etc. will fail. This exception is mainly for us and won't be returned to the user. Not sure what are we considered as a best practice here..

Reply:
I'd recommend just logging an error. Also, you could simplify this as follows:
```kotlin
config.environments.find { it.name == pin.targetEnvironment }
  .also { if (it == null) log.error(...) }
  ?.sendSlackMessage(...)
```
That way you don't have to mess with exception handling in the first place.